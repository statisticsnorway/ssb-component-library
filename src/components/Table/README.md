# Table

> A table component to display data in a structured, grid-like format

### Usage

#### HTML

```html
<div class="ssb-table-wrapper">
  <table class="ssb-table">
    <caption>
      <div class="caption-wrapper">
        <div class="caption-text-wrapper">Caption text</div>
      </div>
    </caption>
    <thead>
      <tr>
        <th>Header</th>
        <th class="align-right">Header<sup>1</sup></th>
        <th></th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>Example</td>
        <td class="align-right">1000</td>
        <td class="align-center">Center value</td>
      </tr>
      <tr>
        <td>Example</td>
        <td class="align-right">500</td>
        <td class="align-center">Center value</td>
      </tr>
    </tbody>
    <tfoot>
      <tr>
        <td colspan="3"><sup>1</sup> Footer text</td>
      </tr>
    </tfoot>
  </table>
</div>
```

#### React

```jsx harmony
<Table caption='Caption text'>
  <TableHead>
    <TableRow>
      <TableCell type='th'>Header</TableCell>
      <TableCell type='th' align='right'>
        Header<sup>1</sup>
      </TableCell>
      <TableCell type='th'></TableCell>
    </TableRow>
  </TableHead>
  <TableBody>
    <TableRow>
      <TableCell>Example</TableCell>
      <TableCell align='right'>1000</TableCell>
      <TableCell align='center'>Center value</TableCell>
    </TableRow>
    <TableRow>
      <TableCell>Example</TableCell>
      <TableCell align='right'>500</TableCell>
      <TableCell align='center'>Center value</TableCell>
    </TableRow>
  </TableBody>
  <TableFooter>
    <TableRow>
      <TableCell type='td' colSpan={3}>
        <sup>1</sup> Footer text
      </TableCell>
    </TableRow>
  </TableFooter>
</Table>
```

Available props:

##### Table

| Name         | Type          | Description                                                      |
| ------------ | ------------- | ---------------------------------------------------------------- |
| className    | string        | Optional styling classes                                         |
| caption      | string        | Optional caption text                                            |
| dataNoteRefs | string        | Optional caption text for data-noterefs attribute                |
| children     | Required node | Table content (TableHead, TableBody, and TableFooter components) |

##### TableHead, TableBody, TableFooter, and TableRow

| Name      | Type          | Description              |
| --------- | ------------- | ------------------------ |
| className | string        | Optional styling classes |
| children  | Required node | Component content        |

##### Table Cell

| Name             | Type                                    | Description                                                                                   |
| ---------------- | --------------------------------------- | --------------------------------------------------------------------------------------------- |
| id               | string                                  | Optional id for the th element. Used by headers attribute etc                                 |
| className        | string                                  | Optional styling classes                                                                      |
| type             | 'th' or 'td'                            | Define what type of element your table cell is. Optional and defaults to td                   |
| rowSpan          | number                                  | Define how many rows the table cell spans. Optional                                           |
| colSpan          | number                                  | Define how many columns the table cell spans. Optional                                        |
| children         | Optional node, string, or number        | Table cell value                                                                              |
| headers          | string                                  | id attributes that provide the headers for the cell. Optional                                 |
| scope            | 'col', 'colgroup', 'row', or 'rowgroup' | Define the cells that the th element relates to. Optional                                     |
| align            | 'left', 'center', or 'right'            | Optional lign table cell content to left, center, or right side                               |
| indentationLevel | '1', '2', or '3'                        | Optional indentation to cell value. Only applies to table content cells that are left aligned |
