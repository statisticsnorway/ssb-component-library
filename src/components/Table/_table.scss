@use '../../style/variables' as variables;

.ssb-table-wrapper {
  width: 100%;
  overflow-x: auto;

  .ssb-table {
    @include roboto;
    line-height: 1.75rem;
    min-width: $desktop-screen; // We need a fixed width for overflow-x to work
    border-collapse: collapse;
    border: 1px solid variables.$ssb-dark-5;

    caption,
    th,
    td {
      border: 1px solid variables.$ssb-dark-5;
    }

    caption,
    td {
      border-bottom: transparent;
    }

    tfoot td > * {
      @media #{$desktop} {
        max-width: 780px;
      }
    }

    .caption-text-wrapper{
      @media #{$desktop} {
        max-width: 998px;
      }
    }

    th,
    td {
      text-align: left;
      padding: 0.5rem 1rem;

      &.level1 {
        padding-left: 1.75rem;
      }

      &.level2 {
        padding-left: 2.5rem;
      }

      &.level3 {
        padding-left: 3.25rem;
      }

      th {
        vertical-align: bottom;
      }
    }

    caption {
      text-align: left;
      position: relative;
      padding: 1.25rem 1rem;

      @media (max-width: $desktop-screen) {
        padding-top: 47px;

        .caption-wrapper .scroll-icon-wrapper {
          margin-top: -100px;
        }

        .caption-wrapper.single-line .scroll-icon-wrapper {
          margin-top: -70px;
        }
      }

      .icon-container {
        padding: 0 1rem;
      }

      .caption-wrapper {
        display: flex;
        justify-content: space-between;
        position: relative;

        .caption-text-wrapper {
          font-weight: bold;
          text-align: left;
          width: 998px;
          padding-top: 0.5rem;
          padding-bottom: 0.5rem;
        }

        .scroll-icon-wrapper {
          display: flex;
          align-items: center;
          position: sticky;
          right: 40px;
          left: 40px;
          visibility: hidden;

          &.visible {
            visibility: visible;
          }
        }

        .scroll-icon {
          cursor: pointer;
          display: flex;
          justify-content: center;
          align-items: center;
          width: 44px;
          height: 44px;
          overflow: hidden;
          background: transparent;

          &:focus {
            svg {
              fill: variables.$ssb-green-4;
              transition: fill 0.18s;
            }

            svg line,
            svg polyline {
              stroke: variables.$ssb-white;
            }

            svg circle {
              stroke: variables.$ssb-green-4;
              transition: stroke 0.16s;
            }
          }

          &:hover {
            svg {
              fill: variables.$ssb-green-4;
              transition: fill 0.18s;
            }

            svg line,
            svg polyline {
              stroke: variables.$ssb-white;
            }

            svg circle {
              stroke: variables.$ssb-green-4;
              transition: stroke 0.16s;
            }
          }

          &:active {
            svg {
              fill: variables.$ssb-dark-5;
              transition: fill 0.18s;
            }

            svg line,
            svg polyline {
              stroke: variables.$ssb-white;
            }

            svg circle {
              stroke: variables.$ssb-dark-5;
              transition: stroke 0.16s;
            }
          }
        }

        .scroll-icon svg {
          width: 32px;
          height: 32px;
          color: variables.$ssb-green-4;
          stroke-width: 1.5px;
        }
      }
    }

    thead {
      border-bottom: 2px solid variables.$ssb-dark-5;

      th[colspan] {
        text-align: center;
      }
    }

    tbody {
      th {
        font-weight: normal;
        border-bottom: transparent;
      }

      tr {
        &:not(:first-child) {
          th,
          td {
            border-top-color: variables.$ssb-dark-2;
          }
        }

        &:nth-child(odd) {
          background-color: variables.$ssb-green-1;
        }

        &:hover,
        &:nth-child(odd):hover {
          background-color: variables.$ssb-purple-1;
        }
      }

      td {
        white-space: nowrap;
      }
    }

    tfoot {
      border-top: 1px solid variables.$ssb-dark-5;

      tr {
        &:first-of-type > td {
          padding-top: 1.75rem;
        }

        &:last-of-type > td {
          padding-bottom: 1.75rem;
        }
      }

      td {
        border: none;
        padding: 0 1rem 0.5rem;
        word-wrap: break-word;
      }
    }

    sup {
      font-size: xx-small;
    }

    .align-right {
      text-align: right;
    }

    .align-left {
      text-align: left;
    }

    .align-center {
      text-align: center;
    }
  }
}

.scroll-icon-active,
.scroll-icon:active {
  svg {
    fill: var(--ssb-dark-5);
    transition: fill 0.18s;
  }

  svg line,
  svg polyline {
    stroke: var(--ssb-white);
  }

  svg circle {
    stroke: var(--ssb-dark-5);
    transition: stroke 0.16s;
  }
}