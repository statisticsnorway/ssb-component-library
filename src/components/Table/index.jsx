import React from 'react';
import PropTypes from 'prop-types';

// TODO:
// * React hooks e.g. useRef
// * WAI-ARIA

const Table = ({ tableClass, caption, thead, tbody, tfoot }) => (
	<div className="ssb-table">
		<table className={`${tableClass ?? ''}`}>
			{caption ? (
				<caption>
					<div className="caption-text-wrapper">
						{caption}
					</div>
				</caption>
			) : null}
			{thead?.length ? (
				<thead>
					{thead.map(() => (
						<>
							<tr>
								<th rowSpan={2} />
								<th colSpan={2}>Header 1</th>
							</tr>
							<tr>
								<th>Header 2</th>
								<th>Header 3
								</th>
							</tr>
						</>
					))}
				</thead>
			) : null}
			{tbody?.length ? (
				<tbody>
					{tbody.map(() => (
						<>
							<tr>
								<td>Body 1</td>
								<td>Body 2</td>
								<td>Body 3</td>
							</tr>
							<tr>
								<td>Body 4</td>
								<td>Body 5</td>
								<td>Body 6</td>
							</tr>
						</>
					))}
				</tbody>
			) : null}
			{tfoot?.length ? (
				<tfoot>
					{tfoot.map(() => (
						<tr>
							<td colSpan="100%"><sup>1</sup>&nbsp;Footer 1</td>
						</tr>
					))}
				</tfoot>
			) : null}
		</table>
	</div>
);

Table.defaultProps = {

};

Table.propTypes = {
	tableClass: PropTypes.string,
	caption: PropTypes.string,
	thead: PropTypes.array,
	tbody: PropTypes.array,
	tfoot: PropTypes.array,
};

export default Table;
