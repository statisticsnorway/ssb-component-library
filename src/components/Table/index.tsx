import React, { forwardRef, ReactNode, useEffect, useState, useRef } from 'react'
import { ArrowLeftCircle, ArrowRightCircle } from 'react-feather'

export interface TableElementProps {
  className?: string
  children: ReactNode
}

export interface TableProps extends TableElementProps {
  caption?: string
  dataNoteRefs?: string
}

export interface TableCellProps {
  id?: string
  className?: string
  type?: 'th' | 'td'
  rowSpan?: number
  colSpan?: number
  headers?: string
  scope?: 'col' | 'colgroup' | 'row' | 'rowgroup'
  align?: 'left' | 'center' | 'right'
  indentationLevel?: '1' | '2' | '3'
  children?: ReactNode | string | number
}

const Table = forwardRef<HTMLTableElement, TableProps>(({ className, caption, dataNoteRefs, children }, ref) => {
  const tableWrapperRef = useRef<HTMLDivElement | null>(null)
  const iconWrapperRef = useRef<HTMLDivElement | null>(null)
  const captionRef = useRef<HTMLDivElement | null>(null)
  const [isOverflowing, setIsOverflowing] = useState(false)
  const [isActive, setIsActive] = useState<{ left: boolean; right: boolean }>({ left: false, right: false })

  type Direction = 'left' | 'right'

  const handleScroll = (direction: Direction) => {
    if (tableWrapperRef.current) {
      const scrollAmount = direction === 'left' ? -380 : 380
      tableWrapperRef.current.scrollBy({ left: scrollAmount, behavior: 'smooth' })
    }
  }

  const handleMouseClick = (direction: Direction) => {
    handleScroll(direction)
    if (document.activeElement instanceof HTMLElement) {
      document.activeElement.blur() // Force blur to remove hover styles
    }
  }

  const handleKeyPress = (event: React.KeyboardEvent<HTMLDivElement>, direction: Direction) => {
    if (event.key === 'Enter' || event.key === ' ') {
      event.preventDefault()
      setIsActive((prev) => ({ ...prev, [direction]: true }))
      handleScroll(direction)
      setTimeout(() => {
        setIsActive((prev) => ({ ...prev, [direction]: false }))
      }, 150)
    }
  }

  useEffect(() => {
    const checkOverflow = () => {
      if (tableWrapperRef.current) {
        const hasOverflow = tableWrapperRef.current.scrollWidth > tableWrapperRef.current.clientWidth
        setIsOverflowing(hasOverflow)
      }
    }

    const checkCaptionHeight = () => {
      if (captionRef.current) {
        const computedStyle = window.getComputedStyle(captionRef.current)
        const lineHeight = parseFloat(computedStyle.lineHeight)
        const paddingTop = parseFloat(computedStyle.paddingTop)
        const paddingBottom = parseFloat(computedStyle.paddingBottom)
        const captionHeight = captionRef.current.clientHeight - paddingTop - paddingBottom
        const lines = Math.round(captionHeight / lineHeight)
        const parentElement = captionRef.current.parentElement as HTMLElement | null
        if (parentElement) {
          if (lines > 1) {
            parentElement.classList.remove('single-line')
          } else {
            parentElement.classList.add('single-line')
          }
        }
      }
    }

    checkOverflow()
    checkCaptionHeight()
    window.addEventListener('resize', checkOverflow)

    return () => {
      window.removeEventListener('resize', checkOverflow)
    }
  }, [])

  return (
    <div className={`ssb-table-wrapper${isOverflowing ? ' overflowing' : ''}`} ref={tableWrapperRef}>
      <table className={`ssb-table${className ? ` ${className}` : ''}`} ref={ref}>
        {caption && (
          <caption data-noterefs={dataNoteRefs}>
            <div className='caption-wrapper' style={{ position: 'relative' }}>
              <div className='caption-text-wrapper' ref={captionRef}>
                {caption}
              </div>
              <div className={`scroll-icon-wrapper ${isOverflowing ? 'visible' : ''}`} ref={iconWrapperRef}>
                <div
                  className={`scroll-icon ${isActive.left ? 'scroll-icon-active' : ''}`}
                  role='button'
                  aria-label='Scroll left'
                  tabIndex={isOverflowing ? 0 : -1}
                  onClick={() => handleMouseClick('left')}
                  onKeyDown={(event) => handleKeyPress(event, 'left')}
                >
                  <ArrowLeftCircle />
                </div>
                <div
                  className={`scroll-icon ${isActive.right ? 'scroll-icon-active' : ''}`}
                  role='button'
                  aria-label='Scroll right'
                  tabIndex={isOverflowing ? 0 : -1}
                  onClick={() => handleMouseClick('right')}
                  onKeyDown={(event) => handleKeyPress(event, 'right')}
                >
                  <ArrowRightCircle />
                </div>
              </div>
            </div>
          </caption>
        )}
        {children}
      </table>
    </div>
  )
})

export default Table

export const TableHead = forwardRef<HTMLTableSectionElement, TableElementProps>(({ className, children }, ref) => {
  if (children) {
    return (
      <thead className={className} ref={ref}>
        {children}
      </thead>
    )
  }
  return null
})

export const TableBody = forwardRef<HTMLTableSectionElement, TableElementProps>(({ className, children }, ref) => {
  if (children) {
    return (
      <tbody className={className} ref={ref}>
        {children}
      </tbody>
    )
  }
  return null
})

export const TableFooter = forwardRef<HTMLTableSectionElement, TableElementProps>(({ className, children }, ref) => {
  if (children) {
    return (
      <tfoot className={className} ref={ref}>
        {children}
      </tfoot>
    )
  }
  return null
})

export const TableRow = forwardRef<HTMLTableRowElement, TableElementProps>(({ className, children }, ref) => {
  if (children) {
    return (
      <tr className={className} ref={ref}>
        {children}
      </tr>
    )
  }
  return null
})

export const TableCell = forwardRef<HTMLTableCellElement, TableCellProps>(
  ({ id, className, children, type = 'td', rowSpan, colSpan, scope, headers, align, indentationLevel }, ref) => {
    const tableCellClasses: string = `${className ?? ''}${indentationLevel && (!align || align === 'left') ? ` level${indentationLevel}` : ''}${align ? ` align-${align}` : ''}`

    if (type === 'th') {
      return (
        <th
          id={id}
          className={tableCellClasses}
          colSpan={colSpan}
          rowSpan={rowSpan}
          scope={scope}
          headers={headers}
          ref={ref}
        >
          {children}
        </th>
      )
    }

    if (type === 'td') {
      return (
        <td className={tableCellClasses} colSpan={colSpan} rowSpan={rowSpan} headers={headers} ref={ref}>
          {children}
        </td>
      )
    }
    return null
  }
)
