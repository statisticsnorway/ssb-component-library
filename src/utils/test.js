import fs from 'fs';
import path from 'path';
import { render as rtlRender } from '@testing-library/react';

export * from '@testing-library/react';
export * from '@testing-library/jest-dom';

const styles = fs.readFileSync(path.relative(process.cwd(), './lib/bundle.css'), 'utf8');

export function render(ui, options) {
	const view = rtlRender(ui, options);

	const styleElement = document.createElement('style');
	styleElement.innerHTML = styles;
	document.body.appendChild(styleElement);

	return view;
}
