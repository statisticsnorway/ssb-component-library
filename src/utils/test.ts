import fs from 'fs'
import path from 'path'
import { RenderOptions, RenderResult, render as rtlRender } from '@testing-library/react'
import { ReactNode } from 'react'

export * from '@testing-library/react'

const styles = fs.readFileSync(path.relative(process.cwd(), './lib/bundle.css'), 'utf8')

export function render(ui: ReactNode, options?: RenderOptions): RenderResult {
  const view = rtlRender(ui, options)

  const styleElement = document.createElement('style')
  styleElement.innerHTML = styles
  document.body.appendChild(styleElement)

  return view
}
