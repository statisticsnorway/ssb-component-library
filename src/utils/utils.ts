// utils.ts
const classNames = (...classes: (string | boolean | undefined)[]): string => classes.filter(Boolean).join(' ')

export default classNames
