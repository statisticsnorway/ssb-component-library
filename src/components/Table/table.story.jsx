import React from 'react'
import Table, { TableHead, TableBody, TableFooter, TableRow, TableCell } from './index'
import Accordion from '../Accordion/index'

export default {
  title: 'Table',
  type: Table,
  parameters: {
    layout: 'fullscreen',
  },
}

export const Default = () => (
  <div style={{ width: '740px', padding: '4rem' }}>
    <Table>
      <TableHead>
        <TableRow>
          <TableCell type='th'>Header</TableCell>
          <TableCell type='th' align='right'>
            Header
          </TableCell>
          <TableCell type='th' />
        </TableRow>
      </TableHead>
      <TableBody>
        <TableRow>
          <TableCell>Example</TableCell>
          <TableCell align='right'>1000</TableCell>
          <TableCell align='center'>Center</TableCell>
        </TableRow>
        <TableRow>
          <TableCell>Example</TableCell>
          <TableCell align='right'>500</TableCell>
          <TableCell align='center'>Center</TableCell>
        </TableRow>
      </TableBody>
    </Table>
  </div>
)

export const Statistics = () => (
  <div style={{ padding: '4rem' }}>
    <Table caption='Lorem ipsum dolor sit amet, consectetur adipiscing elit. '>
      <TableHead>
        <TableRow>
          <TableCell type='th' rowSpan={2} align='right'>
            2020
          </TableCell>
          <TableCell type='th' colSpan={2} scope='colgroup'>
            Second level
          </TableCell>
        </TableRow>
        <TableRow>
          <TableCell type='th' scope='col' align='right'>
            First level
          </TableCell>
          <TableCell type='th' scope='col' align='right'>
            First level<sup>1</sup>
          </TableCell>
          <TableCell type='th' scope='col' align='right'>
            First level
          </TableCell>
          <TableCell type='th' scope='col' align='right'>
            First level
          </TableCell>
        </TableRow>
      </TableHead>
      <TableBody>
        <TableRow>
          <TableCell type='th' scope='row'>
            Total (with numbers)
          </TableCell>
          <TableCell align='right'>1 000</TableCell>
          <TableCell align='right'>2 000</TableCell>
          <TableCell align='right'>3 000</TableCell>
          <TableCell align='right'>4 000</TableCell>
        </TableRow>
        <TableRow>
          <TableCell type='th' scope='row' indentationLevel={1}>
            First level
          </TableCell>
          <TableCell align='right'>50</TableCell>
          <TableCell align='right'>100</TableCell>
          <TableCell align='right'>150</TableCell>
          <TableCell align='right'>200</TableCell>
        </TableRow>
        <TableRow>
          <TableCell type='th' scope='row' indentationLevel={2}>
            Second level
          </TableCell>
          <TableCell align='right'>200</TableCell>
          <TableCell align='right'>650</TableCell>
          <TableCell align='right'>800</TableCell>
          <TableCell align='right'>1050</TableCell>
        </TableRow>
        <TableRow>
          <TableCell type='th' scope='row' indentationLevel={3}>
            Third level
          </TableCell>
          <TableCell align='right'>500</TableCell>
          <TableCell align='right'>200</TableCell>
          <TableCell align='right'>700</TableCell>
          <TableCell align='right'>900</TableCell>
        </TableRow>
      </TableBody>
      <TableFooter>
        <TableRow>
          <TableCell type='td' colSpan={3}>
            <div>
              <sup>1</sup> Lorem ipsum dolor sit amet, consectetur adipiscing elit. Mauris lacinia massa ut velit
              pulvinar, a ullamcorper odio pulvinar. Nulla vulputate congue justo a bibendum. Maecenas aliquet volutpat
              urna ac bibendum. Morbi at pharetra massa. Sed ornare diam eleifend, semper ipsum sed, feugiat purus.
            </div>
          </TableCell>
        </TableRow>
        <TableRow>
          <TableCell type='td' colSpan={3}>
            <sup>2</sup> Lorem ipsum dolor sit amet, consectetur adipiscing elit. Cras non pretium felis.
          </TableCell>
        </TableRow>
        <TableRow>
          <TableCell type='td' colSpan={3}>
            <sup>3</sup> Lorem ipsum dolor sit amet, consectetur adipiscing elit.
          </TableCell>
        </TableRow>
      </TableFooter>
    </Table>
  </div>
)
export const AttachmentTableStatistics = () => (
  <div style={{ padding: '4rem' }}>
    <Accordion subHeader='Tabell 1' header='This is a accordion title with a sub-header'>
      <div style={{ width: '100%' }}>
        <Table caption='Lorem ipsum dolor sit amet, consectetur adipiscing elit. '>
          <TableHead>
            <TableRow>
              <TableCell type='th' rowSpan={2} align='right'>
                2020
              </TableCell>
              <TableCell type='th' colSpan={2} scope='colgroup'>
                Second level
              </TableCell>
            </TableRow>
            <TableRow>
              <TableCell type='th' scope='col' align='right'>
                First level
              </TableCell>
              <TableCell type='th' scope='col' align='right'>
                First level<sup>1</sup>
              </TableCell>
              <TableCell type='th' scope='col' align='right'>
                First level
              </TableCell>
              <TableCell type='th' scope='col' align='right'>
                First level
              </TableCell>
              <TableCell type='th' scope='col' align='right'>
                First level
              </TableCell>
            </TableRow>
          </TableHead>
          <TableBody>
            <TableRow>
              <TableCell type='th' scope='row'>
                Total (with numbers)
              </TableCell>
              <TableCell align='right'>1 000</TableCell>
              <TableCell align='right'>2 000</TableCell>
              <TableCell align='right'>3 000</TableCell>
              <TableCell align='right'>4 000</TableCell>
              <TableCell align='right'>4 000</TableCell>
            </TableRow>
            <TableRow>
              <TableCell type='th' scope='row' indentationLevel={1}>
                First level
              </TableCell>
              <TableCell align='right'>50</TableCell>
              <TableCell align='right'>100</TableCell>
              <TableCell align='right'>150</TableCell>
              <TableCell align='right'>200</TableCell>
              <TableCell align='right'>200</TableCell>
            </TableRow>
            <TableRow>
              <TableCell type='th' scope='row' indentationLevel={2}>
                Second level
              </TableCell>
              <TableCell align='right'>200</TableCell>
              <TableCell align='right'>650</TableCell>
              <TableCell align='right'>800</TableCell>
              <TableCell align='right'>1050</TableCell>
              <TableCell align='right'>1050</TableCell>
            </TableRow>
            <TableRow>
              <TableCell type='th' scope='row' indentationLevel={3}>
                Third level
              </TableCell>
              <TableCell align='right'>500</TableCell>
              <TableCell align='right'>200</TableCell>
              <TableCell align='right'>700</TableCell>
              <TableCell align='right'>900</TableCell>
              <TableCell align='right'>900</TableCell>
            </TableRow>
          </TableBody>
          <TableFooter>
            <TableRow>
              <TableCell type='td' colSpan={3}>
                <div>
                  <sup>1</sup> Lorem ipsum dolor sit amet, consectetur adipiscing elit. Mauris lacinia massa ut velit
                  pulvinar, a ullamcorper odio pulvinar. Nulla vulputate congue justo a bibendum. Maecenas aliquet
                  volutpat urna ac bibendum. Morbi at pharetra massa. Sed ornare diam eleifend, semper ipsum sed,
                  feugiat purus.
                </div>
              </TableCell>
            </TableRow>
            <TableRow>
              <TableCell type='td' colSpan={3}>
                <sup>2</sup> Lorem ipsum dolor sit amet, consectetur adipiscing elit. Cras non pretium felis.
              </TableCell>
            </TableRow>
            <TableRow>
              <TableCell type='td' colSpan={3}>
                <sup>3</sup> Lorem ipsum dolor sit amet, consectetur adipiscing elit.
              </TableCell>
            </TableRow>
          </TableFooter>
        </Table>
      </div>
    </Accordion>
  </div>
)
