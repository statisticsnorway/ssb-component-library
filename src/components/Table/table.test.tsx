import React from 'react'
import { render } from '../../utils/test'
import Table, { TableHead, TableBody, TableFooter, TableRow, TableCell } from './index'

describe('Render Table component', () => {
  test('Matches the snapshot', () => {
    const { asFragment } = render(
      <Table caption='Table caption'>
        <TableHead>
          <TableRow>
            <TableCell type='th' rowSpan={2} align='right'>
              Header with footnote <sup>1</sup>
            </TableCell>
            <TableCell type='th' scope='colgroup'>
              Alignment
            </TableCell>
          </TableRow>
          <TableRow>
            <TableCell type='th' scope='col' align='right'>
              Header right
            </TableCell>
            <TableCell type='th' scope='col' align='center'>
              Header center
            </TableCell>
          </TableRow>
        </TableHead>
        <TableBody>
          <TableRow>
            <TableCell type='th' scope='row'>
              Value
            </TableCell>
            <TableCell align='right'>500</TableCell>
            <TableCell align='center'>Center</TableCell>
          </TableRow>
        </TableBody>
        <TableFooter>
          <TableRow>
            <TableCell colSpan={3}>
              <sup>1</sup> Footnote description
            </TableCell>
          </TableRow>
        </TableFooter>
      </Table>
    )
    expect(asFragment()).toMatchSnapshot()
  })

  test('Simple table without caption and footer', () => {
    const { asFragment } = render(
      <Table>
        <TableHead>
          <TableRow>
            <TableCell type='th'>Header</TableCell>
          </TableRow>
        </TableHead>
        <TableBody>
          <TableRow>
            <TableCell>Value</TableCell>
          </TableRow>
        </TableBody>
      </Table>
    )
    expect(asFragment()).toMatchSnapshot()
  })
})
