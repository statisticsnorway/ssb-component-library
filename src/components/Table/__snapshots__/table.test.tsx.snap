// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Table component Matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="ssb-table-wrapper"
  >
    <table
      class="ssb-table"
    >
      <caption>
        <div
          class="caption-wrapper single-line"
          style="position: relative;"
        >
          <div
            class="caption-text-wrapper"
          >
            Table caption
          </div>
          <div
            class="scroll-icon-wrapper "
          >
            <div
              aria-label="Scroll left"
              class="scroll-icon "
              role="button"
              tabindex="-1"
            >
              <svg
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <circle
                  cx="12"
                  cy="12"
                  r="10"
                />
                <polyline
                  points="12 8 8 12 12 16"
                />
                <line
                  x1="16"
                  x2="8"
                  y1="12"
                  y2="12"
                />
              </svg>
            </div>
            <div
              aria-label="Scroll right"
              class="scroll-icon "
              role="button"
              tabindex="-1"
            >
              <svg
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <circle
                  cx="12"
                  cy="12"
                  r="10"
                />
                <polyline
                  points="12 16 16 12 12 8"
                />
                <line
                  x1="8"
                  x2="16"
                  y1="12"
                  y2="12"
                />
              </svg>
            </div>
          </div>
        </div>
      </caption>
      <thead>
        <tr>
          <th
            class=" align-right"
            rowspan="2"
          >
            Header with footnote 
            <sup>
              1
            </sup>
          </th>
          <th
            class=""
            scope="colgroup"
          >
            Alignment
          </th>
        </tr>
        <tr>
          <th
            class=" align-right"
            scope="col"
          >
            Header right
          </th>
          <th
            class=" align-center"
            scope="col"
          >
            Header center
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <th
            class=""
            scope="row"
          >
            Value
          </th>
          <td
            class=" align-right"
          >
            500
          </td>
          <td
            class=" align-center"
          >
            Center
          </td>
        </tr>
      </tbody>
      <tfoot>
        <tr>
          <td
            class=""
            colspan="3"
          >
            <sup>
              1
            </sup>
             Footnote description
          </td>
        </tr>
      </tfoot>
    </table>
  </div>
</DocumentFragment>
`;

exports[`Render Table component Simple table without caption and footer 1`] = `
<DocumentFragment>
  <div
    class="ssb-table-wrapper"
  >
    <table
      class="ssb-table"
    >
      <thead>
        <tr>
          <th
            class=""
          >
            Header
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td
            class=""
          >
            Value
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;
